### ISSUE TITLE
 You already added the platform "debian-openssl-1.1.x" to the "generator" block in the "schema.prisma" file as described in https://pris.ly/d/client-generator, but something went wrong. 

### ISSUE THREAD

**Issue by satvikpendem** [prisma/prisma#15631]
I'm getting an error on a Docker `node:slim` container:

```json
{
  "errors": [
    {
      "message": "\nInvalid `prisma.user.findUnique()` invocation:\n\n\nQuery engine library for current platform \"debian-openssl-1.1.x\" could not be found.\nYou incorrectly pinned it to debian-openssl-1.1.x\n\nThis probably happens, because you built Prisma Client on a different platform.\n(Prisma Client looked in \"/usr/src/app/generated/prisma-client/runtime/libquery_engine-debian-openssl-1.1.x.so.node\")\n\nSearched Locations:\n\n  /usr/src/app/.prisma/client\n  /usr/src/app/src/generated/prisma-client\n  /usr/src/app/generated/prisma-client\n  /usr/src/app/generated/prisma-client\n  /usr/src/app/prisma\n  /tmp/prisma-engines\n  /usr/src/app/generated/prisma-client\n\nYou already added the platform \"debian-openssl-1.1.x\" to the \"generator\" block\nin the \"schema.prisma\" file as described in https://pris.ly/d/client-generator,\nbut something went wrong. That's suboptimal.\n\nPlease create an issue at https://github.com/prisma/prisma/issues/new",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "posts"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "clientVersion": "4.4.0"
        }
      }
    }
  ],
  "data": {
    ...
  }
}
```

My schema:

```
generator client {
    provider      = "prisma-client-js"
    output        = "../generated/prisma-client"
    binaryTargets = ["native"]
}

generator pothos {
    provider      = "prisma-pothos-types"
    output        = "../generated/pothos-types.ts"
    binaryTargets = ["native"] //, "debian-openssl-1.1.x"]
}

datasource db {
    provider = "postgresql"
    url      = env("DATABASE_URL")
}

model User {
    id        Int      @id @default(autoincrement())
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt
    username  String   @unique
    email     String   @unique
    password  String
    posts     Post[]
}

model Post {
    id        Int      @id @default(autoincrement())
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt
    title     String
    content   String?
    author    User     @relation(fields: [authorId], references: [id])
    authorId  Int
    points    Int      @default(0)
}
```

**Comment by annibuliful** [comment]
I had been gotten the same issue for long time ago
could you try to change `node:slim` to `node:16`?

This is my prisma config
```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["interactiveTransactions"]
  binaryTargets   = ["native", "linux-musl"]
}

```

**Comment by annibuliful** [comment]
my full Docker file. please skip prisma generate because I have 3 databases instances 😄 
```
FROM node:16
WORKDIR /app

COPY ./dist/apps/auth-backend .
COPY prisma ./prisma/ 
COPY .env .

EXPOSE 3001

RUN apt-get update
RUN apt-get install -y openssl
RUN npm install -g pnpm

RUN pnpm install
RUN pnpm install pino-pretty

RUN npx prisma generate --schema prisma/primary/schema.prisma
RUN npx prisma generate --schema prisma/job/schema.prisma
RUN npx prisma generate --schema prisma/internal/schema.prisma

CMD node ./main.js
```

**Comment by satvikpendem** [comment]
What fixed it for me was using a single build container for Docker rather than a multi-step builds. I'm not sure why exactly it helped but it might have been something to do with how Prisma detects differences between builds.

**Comment by annibuliful** [comment]
Maybe related to base os binary is different when you build with multiple step buils.

**Comment by millsp** [comment]
Hey @satvikpendem would you able to share the exact steps for us to reproduce this? Only if you have time, we'd appreciate a reproduction. That would be very helpful. Thanks @annibuliful for the extra information!

**Comment by satvikpendem** [comment]
I'm not sure exactly what caused the issue in the first place so I'm not sure how to reproduce. What I do know is that I was using a multistep Docker build.

1. I have a `src` folder inside which I have the `prisma` and `generated` folders. This is because I wanted to create a fully contained `dist` folder for production.
2. I run `prisma generate` which generates a new `generated` folder.
3. In `src` I am using `swc` to compile my TypeScript files into JS into the `dist` folder.
4. `swc` doesn't copy non-TS files so I `cp src/prisma src/generated dist`.
5. When I try to run `node dist/index.js` Prisma complains as above.

I believe what happened was Prisma somehow knows it was copied to dist and throws an error.

**Comment by millsp** [comment]
Thanks for the detail. Have you also copied the engine file near your new bundle?

**Comment by millsp** [comment]
Thanks. So I guess this did not only happen in docker, but also in your dev environment?
I think I can see why (in the code) why the engine would not be found in such a case.

**Comment by satvikpendem** [comment]
Hm I am actually generating it in a custom location already:

```
generator client {
    provider      = "prisma-client-js"
    output        = "../generated/prisma-client" // this becomes src/generated/prisma-client
    binaryTargets = ["native"]
}
```

And yes it happens in my dev env as well as Docker, however it seems to be fixed now that I have copied over the query engine file to `dist`.

**Comment by millsp** [comment]
Hey, we wanted to give you an update about this. We have around 10 issues that we have identified to be in the "engine not found" category. Looking at these, we learned that there are four main categories responsible for these errors:

* Your client was generated with `native` (default) on one platform, but used on a different platform at runtime.
* You pinned `binaryTargets` incorrectly, and the correct engine was therefore not included at runtime.
* A bundler/framework changed the Prisma Client but did not copy the necessary engine for the runtime.
* Deployment/tooling did not copy the engine correctly so the correct engine was not included at runtime.

These cases were previously not detected. On top of that, the error messages were confusing and were neither helpful for you or us. With that, it is likely that many of the cases above overlapped each other. So while we cannot fix a missing engine, we can detect how and why it happened. Thanks to that, we now display more useful errors to help you out of trouble. And in some cases, we will ask some input from you to learn more about your project so we can further improve.

If something is still off for you, feel free to create a new issue.
